[core][pruning][feature] cuSPARSELt kernels and ops

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102133

This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

jcaip committed:
build

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt build integration"


This PR modifies the pytorch build process to support CUSPARSELT.

Currently, cuSPARSELt is built as part of libtorch_cuda.so

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

jcaip committed:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

jcaip commented:
@pytorchbot rebase main

jcaip commented:
@pytorchbot --help

jcaip commented:
@pytorchbot rebase -b viable/strict

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_semi_structured.py:
```diff
@@ -33,10 +33,22 @@
+    SEMI_STRUCTURED_SUPPORTED_BACKENDS.append("cutlass")
+
+    # check if cslt is available for now using this:
+    # TODO when we add cusparselt as a backend, we can update this to be use torch.cusparselt.is_available()
```
What's stopping you from doing that?

jcaip commented:
@pytorchbot merge

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

Commit:
Update on "[core][pruning][feature] cuSPARSELt kernels and ops"


This PR contains two new private ops, added for cuSPARSELt support.

These ops call into the cuSPASRELt kernels using the bindings they
provide. For more information, see the documentation
[here](https://docs.nvidia.com/cuda/cusparselt/index.html).

The two new private ops added are:
```
_cslt_compress()
_cslt_sparse_mm()
```

_cslt_compress is an op that reuturns the compressesed matrix given a
sparse matrix that is passed in.

_cslt_sparse_mm is an op that expects a compressed matrix (the result of
_cslt_compress) and a dense matrix and performs sparse-dense matmul

These ops will throw runtime errors if they cusparselt is not present.

This PR also modifies the test and tensor sublass to reflect the new
cuSPARSELt support.

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "breaking internal lots of builds, see D48144534" -c ghfirst

jcaip commented:
@pytorchbot revert -m "breaking lots of internal builds, see D48144534" -c ghfirst
